### PR TITLE
[InstCombine] fold (Binop phi(a, b) phi(b, a)) -> (Binop a, b) while Binop is commutative.

### LINKED ISSUE #73905: [InstCombine] Missing optimization : fold `mul (phi a, b), (phi b, a)` to `mul a, b`
Similar to #73904

Alive2 proof: https://alive2.llvm.org/ce/z/RDGBSJ
Missed example: https://godbolt.org/z/9qK8G8o16

### Description:
```llvm
define i8 @src(i1 %c, i8 %a, i8 %b)  {
entry:
  br i1 %c, label %then, label %end
then:
  call void @dummy()
  br label %end
end:
  %phi1 = phi i8 [%a, %entry], [%b, %then]
  %phi2 = phi i8 [%b, %entry], [%a, %then]
  %mul = mul i8 %phi1, %phi2
  ret i8 %mul
}
```
can be folded to:
```llvm
define i8 @tgt(i1 %c, i8 %a, i8 %b) {
entry:
  br i1 %c, label %then, label %end
then:
  call void @dummy()
  br label %end
end:
  %mul = mul i8 %a, %b
  ret i8 %mul
}
```

### Real-world motivation

This snippet of IR is derived from [src/redis/evict.c](https://github.com/redis/redis/blob/unstable/src/evict.c) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by sun-jacobi**
This patch closes #73905

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1101]:
```suggestion
  if (LHS->getNumIncomingValues() != 2 || RHS->getNumIncomingValues() != 2)
```
It would be clearer than `getNumOperands`.

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:4246]:
```suggestion
  assert(II.isCommutative() && "Instruction should be commutative");
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1124]:
```suggestion
  assert(I.isCommutative() && "Instruction should be commutative");
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1136]: Could you please add some tests with IR flag propagation (e.g., nowrap flags and fast-math flags)?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1508]: Call this from foldBinopWithPhiOperands() instead of repeating it everywhere?

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1110]: Please check that the phi nodes are in the same basic block instead.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1101]: I don't think we really need this restriction. We just need all pairs of phi operands to have the same two values (in either order). The number of pairs doesn't matter.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (sun-jacobi) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1110]: Does this mean we need to add a check to see if the phi nodes are in the same basic block?

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1110]: Yes, check that `LHS->getParent() == RHS->getParent()`.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (tschuett) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1131]: Do you need to restrict to 2 incoming values per phi here? matchSymmetricPhiNodesPair could return true for phis with 10^10 incoming values.

Comment (sun-jacobi) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1131]: Could you explain more ? I think current`matchSymmetricPhiNodesPair` returns false for 2 incoming values per phi case.

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:4256]:
```suggestion
    replaceOperand(II, 1, RHS->getIncomingValue(0));
```
I think with the updated implementation we should do this?

Though I think it would be safer to return something like `std::optional<std::pair<Value *, Value *>>` from `matchSymmetricPhiNodesPair()`, which avoids any implicit contract between these functions.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1104]: No need to check this, it's implied by same parent.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1111]: No need to check this, it's implied by same parent.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1120]: Same here.

File: llvm/lib/Transforms/InstCombine/InstCombineInternal.h
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineInternal.h:288]:
```suggestion
  // while op is a commutative intrinsic call.
```

File: llvm/lib/Transforms/InstCombine/InstCombineInternal.h
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineInternal.h:281]:
```suggestion
  // Match a pair of Phi Nodes like
```

File: llvm/lib/Transforms/InstCombine/InstCombineInternal.h
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineInternal.h:283]:
```suggestion
  // Return the matched two operands.
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1110]:
```suggestion
  Value *L0 = LHS->getIncomingValueForBlock(B0);
  Value *R0 = RHS->getIncomingValueForBlock(B0);
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1117]:
```suggestion
    Value *L1 = LHS->getIncomingValueForBlock(B1);
    Value *R1 = RHS->getIncomingValueForBlock(B1);
    if ((L0 == L1 && R0 == R1) || (L0 == R1 && R0 == L1))
      continue;
```

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1109]: Using getIncomingValueForBlock() is inefficient for large phis (the function will be quadratic).

What you can do is add a check `if (!equal(LHS->blocks(), RHS->blocks()) return false;` and then just work on the incoming values ignoring the blocks.

InstCombine canonicalizes the block order for phis in the same block, so this check only guards against phi nodes that have no been canonicalized yet -- it will eventually succeed.

**Review by nikic — COMMENTED** [review]
Some more nits.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (sun-jacobi) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1109]: Is there any efficient way to `equal(LHS->blocks(), RHS->blocks())` ?

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1109]: Doesn't writing literally that work?

Comment (sun-jacobi) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: Oh, it works. Thanks a lot !

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
